src/interface/blas2_interface.hpp

```diff
@@ -733,16 +733,73 @@ typename sb_handle_t::event_t _spr_impl(sb_handle_t& sb_handle, char _Uplo,
   const index_t globalSize = localSize * nWGPerCol;
 
   if (Upper) {
-    auto spr = make_gerp<true, true>(mA, _N, _alpha, vx, _incx, vx, _incx);
+    auto spr = make_spr<true, true>(mA, _N, _alpha, vx, _incx, vx, _incx);
     return ret = concatenate_vectors(
                ret, sb_handle.execute(spr, localSize, globalSize));
   } else {
-    auto spr = make_gerp<true, false>(mA, _N, _alpha, vx, _incx, vx, _incx);
+    auto spr = make_spr<true, false>(mA, _N, _alpha, vx, _incx, vx, _incx);
     return ret = concatenate_vectors(
                ret, sb_handle.execute(spr, localSize, globalSize));
   }
 }
 
+/*! _SPR2.
+ * @brief Implementation of the rank 2 operation
+ */
+/*
+sspr2 	( 	character  	UPLO,
+   integer  	N,
+   real  	ALPHA,
+   real, dimension(N)  	X,
+   integer  	INCX,
+   real, dimension(N)  	Y,
+   integer  	INCY,
+   real, dimension(N, N + 1 / 2)  	AP
+ )
+*/
+template <typename sb_handle_t, typename index_t, typename element_t,
+          typename container_t0, typename increment_t, typename container_t1,
+          typename container_t2>
+typename sb_handle_t::event_t _spr2_impl(sb_handle_t& sb_handle, char _Uplo,
+                                         index_t _N, element_t _alpha,
+                                         container_t0 _vx, increment_t _incx,
+                                         container_t1 _vy, increment_t _incy,
+                                         container_t2 _mPA) {
+  // throw exception if invalid arguments
+  if (_N <= 0) {
+    throw std::invalid_argument("Invalid vector size");
+  }
+
+  // bail out early if alpha == 0
+  if (_alpha == (element_t)0) {
+    typename sb_handle_t::event_t event;
+    return event;
+  }
+
+  typename sb_handle_t::event_t ret;
+  _Uplo = tolower(_Uplo);
+  const int Upper = _Uplo == 'u';
+  auto mA = make_matrix_view<col_major>(_mPA, _N, (_N + 1) / 2, _N);
+  auto vx = make_vector_view(_vx, _incx, (1 + (_N - 1) * std::abs(_incx)));
+  auto vy = make_vector_view(_vy, _incy, (1 + (_N - 1) * std::abs(_incy)));
```

Comment:
I think the 3rd argument is the number of vector elements, not the total size of the buffer.

```suggestion
  auto vx = make_vector_view(_vx, _incx, _N);
  auto vy = make_vector_view(_vy, _incy, _N);
```